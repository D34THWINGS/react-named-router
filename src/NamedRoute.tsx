import React from 'react';
import { Route, RouteComponentProps, RouteProps } from 'react-router-dom';
import invariant from 'tiny-invariant';

import { NamedRouteConfig, useNamedRouting } from './NamedRouter';

export interface NamedRouteProps<TParams = object> extends Omit<RouteProps, 'path'> {
  name: string;
  children?: React.ReactNode;
}

export interface NamedRouteComponentProps<P, C, S> extends RouteComponentProps<P, C, S> {
  route: NamedRouteConfig;
}

export const NamedRoute: React.FC<NamedRouteProps> = ({
  name,
  children,
  ...otherProps
}) => {
  const context = useNamedRouting();

  invariant(context, 'You should not use <NamedRoute> outside a <NamedRouter>');

  const route = context.getRoute(name);

  return (
    <Route {...route} {...otherProps}>
      {children}
    </Route>
  );
};

NamedRoute.displayName = 'NamedRoute';
