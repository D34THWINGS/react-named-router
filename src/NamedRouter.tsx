import React, { createContext, useContext } from 'react';
import { BrowserRouter, BrowserRouterProps } from 'react-router-dom';
import { Route } from 'react-router';
/* eslint-disable import/no-extraneous-dependencies */
import { RouteConfig } from 'react-router-config';
/* eslint-enable */

import { BaseRoutingContext, buildRoutingContext, RoutingContext } from './utils';

export interface NamedRouteConfig extends RouteConfig {
  name?: string;
  title?: string;
  routes?: NamedRouteConfig[];
}

export interface NamedRouterProps<TRouterProps = BrowserRouterProps> {
  routes: NamedRouteConfig[];
  routerComponent?: React.ComponentType<TRouterProps>;
  routerProps?: TRouterProps;
  children?: React.ReactNode;
}

export const routerContext = createContext<BaseRoutingContext | RoutingContext>(buildRoutingContext([]));
export const useNamedRouting = () => useContext(routerContext);

const NamedRouter = ({
  routerComponent: RouterComponent = BrowserRouter,
  routerProps,
  routes,
  children,
}: NamedRouterProps) => (
  <RouterComponent {...routerProps}>
    <Route>
      {({ history }) => (
        <routerContext.Provider value={buildRoutingContext(routes, history)}>
          {children}
        </routerContext.Provider>
      )}
    </Route>
  </RouterComponent>
);

export default NamedRouter;
