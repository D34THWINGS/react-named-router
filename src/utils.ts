// eslint-disable-next-line import/no-extraneous-dependencies
import { History } from 'history';

import { NamedRouteConfig } from './NamedRouter';

export interface ExtendedRouteConfig extends NamedRouteConfig {
  parents: string[];
  regex?: RegExp;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const buildRoutePath = (map: Map<string, ExtendedRouteConfig>, name: string, params: any = {}) => {
  const route = map.get(name);
  if (!route) {
    throw new Error(`Undefined route "${name}"`);
  }
  if (!route.path) {
    throw new Error(`Route "${name}" does not have a path`);
  }
  return route.path
    .replace(/:\w+\??/g, (match) => {
      const paramKey = match.replace(/[:?]/g, '');
      const paramValue = params[paramKey];
      if (!paramValue && !match.includes('?')) {
        throw new Error(`Missing value for required param "${paramKey}"`);
      }
      return paramValue || '';
    })
    .replace(/\/$/, '');
};

export const mapRoutes = (
  routes: NamedRouteConfig[],
  parents: string[] = [],
  map: Map<string, ExtendedRouteConfig> = new Map(),
) => {
  routes.forEach((route) => {
    if (route.routes) {
      mapRoutes(route.routes, route.name ? [...parents, route.name] : parents, map);
    }

    if (!route.name && route.path) {
      throw new Error(`Route ${route.path} has no name`);
    }

    if (!route.name) {
      return;
    }

    const existingRoute = map.get(route.name);
    if (process.env.NODE_ENV !== 'production' && existingRoute) {
      // eslint-disable-next-line no-console
      console.warn(`Duplicate definition for route ${route.name}:\n- ${existingRoute.path}\n- ${route.path}`);
    }

    map.set(route.name, {
      ...route,
      parents,
      ...(route.path ? {
        regex: new RegExp(
          `^${route.path.replace(/\/:\w+\?/g, '/?([\\w_-]*)').replace(/:\w+/g, '([\\w_-]+)')}\\/?$`,
        ),
      } : {}),
    });
  });

  return map;
};

export class BaseRoutingContext {
  protected readonly routesMap: Map<string, ExtendedRouteConfig>;

  public constructor(routesMap: Map<string, ExtendedRouteConfig>) {
    this.routesMap = routesMap;
  }

  public match = (pathname: string) => Array.from(this.routesMap.values())
    .find(route => !!route.regex && route.regex.test(pathname)) || null;

  public getPath = <TParams>(name: string, params?: TParams) => buildRoutePath(this.routesMap, name, params);

  public getRoute = (name: string) => {
    const route = this.routesMap.get(name);
    if (!route) {
      throw new Error(`Undefined route "${name}"`);
    }
    return route;
  }
}

export class RoutingContext extends BaseRoutingContext {
  protected history: History;

  public constructor(routesMap: Map<string, ExtendedRouteConfig>, history: History) {
    super(routesMap);

    this.history = history;
  }

  public push = <TParams>(name: string, params?: TParams) => this.history
    .push(buildRoutePath(this.routesMap, name, params));

  public replace = <TParams>(name: string, params?: TParams) => this.history
    .replace(buildRoutePath(this.routesMap, name, params));
}

/* eslint-disable import/export */
export function buildRoutingContext(routes: NamedRouteConfig[]): BaseRoutingContext;
export function buildRoutingContext(routes: NamedRouteConfig[], history?: History): RoutingContext;
export function buildRoutingContext(
  routes: NamedRouteConfig[],
  history?: History,
): BaseRoutingContext | RoutingContext {
  const map = mapRoutes(routes);

  if (!history) {
    return new BaseRoutingContext(map);
  }

  return new RoutingContext(map, history);
}
/* eslint-enable import/export */
