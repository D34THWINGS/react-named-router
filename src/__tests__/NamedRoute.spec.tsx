import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { StaticRouter, StaticRouterProps } from 'react-router';
import { NamedRouter } from '../NamedRouter';
import { NamedRoute } from '../NamedRoute';
import { NamedRouteConfig } from '../utils';

Enzyme.configure({ adapter: new Adapter() });

describe('NamedSwitch', () => {
  it('should render only first matching route', () => {
    // Given
    const routerProps: StaticRouterProps = { location: '/foo' };
    const routes: NamedRouteConfig[] = [{
      name: 'foo',
      path: '/foo',
      exact: true,
      component: () => null,
    }];

    // When
    const wrapper = mount((
      <NamedRouter
        routes={routes}
        routerComponent={StaticRouter}
        routerProps={routerProps}
      >
        <NamedRoute name="foo" />
      </NamedRouter>
    ));

    // Then
    expect(wrapper).toMatchSnapshot();
  });
});
