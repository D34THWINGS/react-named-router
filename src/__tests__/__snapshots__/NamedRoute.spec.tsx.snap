// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NamedSwitch should render only first matching route 1`] = `
<NamedRouter
  routerComponent={[Function]}
  routerProps={
    Object {
      "location": "/foo",
    }
  }
  routes={
    Array [
      Object {
        "component": [Function],
        "exact": true,
        "name": "foo",
        "path": "/foo",
      },
    ]
  }
>
  <StaticRouter
    location="/foo"
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/foo",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      staticContext={Object {}}
    >
      <Route>
        <NamedRoute
          name="foo"
        >
          <Route
            component={[Function]}
            exact={true}
            name="foo"
            parents={Array []}
            path="/foo"
            regex={/\\^\\\\/foo\\\\/\\?\\$/}
          >
            <component
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "createHref": [Function],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "pathname": "/foo",
                    "search": "",
                    "state": undefined,
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              location={
                Object {
                  "hash": "",
                  "pathname": "/foo",
                  "search": "",
                  "state": undefined,
                }
              }
              match={
                Object {
                  "isExact": true,
                  "params": Object {},
                  "path": "/foo",
                  "url": "/foo",
                }
              }
              staticContext={Object {}}
            />
          </Route>
        </NamedRoute>
      </Route>
    </Router>
  </StaticRouter>
</NamedRouter>
`;
